internal/scheduler/submitcheck.go

```diff
@@ -1,71 +1,122 @@
+		priorityClasses:    schedulingConfig.Preemption.PriorityClasses,
+		gangIdAnnotation:   schedulingConfig.GangIdAnnotation,
+		executorById:       map[string]minimalExecutor{},
+		priorities:         schedulingConfig.Preemption.AllowedPriorities(),
+		indexedResources:   schedulingConfig.IndexedResources,
+		indexedTaints:      schedulingConfig.IndexedTaints,
+		indexedNodeLabels:  schedulingConfig.IndexedNodeLabels,
+		executorRepository: executorRepository,
+		clock:              clock.RealClock{},
 	}
 }
 
-// RegisterNodeDb adds a NodeDb to use when checking if a pod can be scheduled.
-// To only check static scheduling requirements, set NodeDb.CheckOnlyStaticRequirements = true
-// before registering it.
-func (srv *SubmitChecker) RegisterNodeDb(executor string, nodeDb *NodeDb) {
-	srv.mu.Lock()
-	defer srv.mu.Unlock()
-	if srv.nodeDbByExecutor == nil {
-		srv.nodeDbByExecutor = make(map[string]*NodeDb)
+func (srv *SubmitChecker) Run(ctx context.Context) error {
+	srv.updateExecutors(ctx)
+	ticker := time.NewTicker(1 * time.Minute)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			srv.updateExecutors(ctx)
+		}
+	}
+}
+
+func (srv *SubmitChecker) updateExecutors(ctx context.Context) {
+	executors, err := srv.executorRepository.GetExecutors(ctx)
+	if err != nil {
+		log.WithError(err).Error("Error fetching executors")
+		return
+	}
+	for _, executor := range executors {
+		nodeDb, err := srv.constructNodeDb(executor.Nodes)
+		if err == nil {
+			srv.mu.Lock()
+			srv.executorById[executor.Id] = minimalExecutor{
+				nodeDb:     nodeDb,
+				updateTime: executor.LastUpdateTime,
+			}
+			srv.mu.Unlock()
+			if err != nil {
+				log.WithError(err).Errorf("Error constructing node db for executor %s", executor.Id)
+			}
+		} else {
+			log.WithError(err).Warnf("Error clearing nodedb for executor %s", executor.Id)
+		}
+
 	}
-	srv.nodeDbByExecutor[executor] = nodeDb
 }
 
 func (srv *SubmitChecker) CheckApiJobs(jobs []*api.Job) (bool, string) {
-	// // First, check if all jobs can be scheduled individually.
-	// // TODO: Disabled for now; the SubmitChecker would reject all jobs until populated.
-	// for i, job := range jobs {
-	// 	reqs := PodRequirementsFromJob(srv.priorityClasses, job)
-	// 	canSchedule, reason := srv.Check(reqs)
-	// 	if !canSchedule {
-	// 		return canSchedule, fmt.Sprintf("%d-th job unschedulable:\n%s", i, reason)
-	// 	}
-	// }
+	// First, check if all jobs can be scheduled individually.
+	for i, job := range jobs {
+		reqs := PodRequirementsFromLegacySchedulerJob(job, srv.priorityClasses)
+		canSchedule, reason := srv.check([]*schedulerobjects.PodRequirements{reqs})
+		if !canSchedule {
+			return canSchedule, fmt.Sprintf("%d-th job unschedulable:\n%s", i, reason)
+		}
+	}
 	// Then, check if all gangs can be scheduled.
-	for gangId, jobs := range groupJobsByAnnotation(srv.gangIdAnnotation, jobs) {
+	for gangId, jobs := range GroupJobsByAnnotation(srv.gangIdAnnotation, jobs) {
```

Comment:
You could use `GroupJobsByAnnotation` to group by gang id in `submit_to_log.go`. Avoid duplicating some code.

Reply:
this function is different- the function there groups all jobs without an annotation into a single "gang" with no id which subsequent code in that file knows to ignore.  The function here groups all jobs without an annotation into a gang of 1.